baselines/ppo2/policies.py

```diff
@@ -0,0 +1,207 @@
+        X = observ_placeholder if observ_placeholder is not None else observation_placeholder(ob_space,
+                                                                                              batch_size=nbatch)
+        dones = tf.placeholder_with_default(np.zeros([X.shape[0]]), [X.shape[0]], name='dones')
+
+        encoded_x = encode_observation(ob_space, X)
+
+        with tf.variable_scope('load_rnn_memory'):
+            if is_rnn_network(policy_network):
+                policy_state, policy_network_ = policy_network(encoded_x, dones)
+            else:
+                policy_network_ = policy_network
+
+            if value_network == 'shared':
+                value_network_ = value_network
+            else:
+                if value_network == 'copy':
+                    value_network_ = policy_network
+                else:
+                    assert callable(value_network)
+                    value_network_ = value_network
+
+                if is_rnn_network(value_network_):
+                    value_state, value_network_ = value_network_(encoded_x, dones)
+
+            if policy_state or value_state:
+                states_list = [state for state in [policy_state, value_state] if state]
+                states = tf.concat(states_list, axis=1)
+                state_placeholder = tf.placeholder_with_default(states, states.get_shape())
+                index = 0
+                for state in states_list:
+                    assert state.get_shape().ndims == 2
+                    size = int(state.get_shape()[1])
+                    state_map[state] = state_placeholder[:, index:index + size]
+                    index += size
+
+        with tf.variable_scope('policy_latent', reuse=tf.AUTO_REUSE):
+            if is_rnn_network(policy_network_):
+                policy_latent, next_policy_state = \
+                    policy_network_(encoded_x, dones, state_map[policy_state])
+                next_states_list.append(next_policy_state)
+            else:
+                policy_latent = policy_network_(encoded_x)
+
+        with tf.variable_scope('value_latent', reuse=tf.AUTO_REUSE):
+            if value_network_ == 'shared':
+                value_latent = policy_latent
+            elif is_rnn_network(value_network_):
+                value_latent, next_value_state = \
+                    value_network_(encoded_x, dones, state_map[value_state])
+                next_states_list.append(next_value_state)
+            else:
+                value_latent = value_network_(encoded_x)
+
+        with tf.name_scope("save_rnn_memory"):
+            if policy_state or value_state:
+                next_states = tf.concat(next_states_list, axis=1)
+                state_info = {'current': state_placeholder,
+                              'next': next_states, }
+                update_op = []
+                index = 0
+                for state in [state for state in [policy_state, value_state] if state]:
```

Comment:
I think this can be expressed in a somewhat more readable (if maybe less concise) way as 
```
for state in [policy_state, value_state]:
    if state: 
        ...
```

Reply:
also, coming to think about it, even more readable version would be something like
```
for state, next_state in zip(state_list, next_state_list):
    if state and next_state:
       update_op.append(state.assign(next_state))
```
(in other words, I am not a fan of concatenating next_states and then slicing them again)

Reply:
You're right. I going to apply the code.

Reply:
removed at b6e6c52